### PR TITLE
fixed 'System.ObjectDisposedException' when closing/disposing connection

### PR DESCRIPTION
**Pull request by nazmul1985**
## Proposed Changes
Lock _heartbeatWriteTimer and _heartbeatReadTimer in Connection.cs before disposing and inside the timer elapsed delegate to avoid the following exception:
   ```
System.ObjectDisposedException occurred
Message: Exception thrown: 'System.ObjectDisposedException' in mscorlib.dll
Additional information: Cannot access a disposed object.
   at System.Threading.TimerQueueTimer.Change(UInt32 dueTime, UInt32 period)
```
Please see [Issue#220](https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/220) for details.
## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x ] Bugfix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x ] I have read the `CONTRIBUTING.md` document
- [ x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

**Comment by michaelklishin** [comment]
Do you have any evidence that the exception is due to a concurrency hazard? If so, what is it?

File: projects/client/RabbitMQ.Client/src/client/impl/Connection.cs [hunk 1]
```diff
@@ -1040,99 +1042,127 @@ public void StartMainLoop(bool useBackgroundThread)
 #endif
         }
 
+        private readonly object heartBeatReadLock = new object();
+        private readonly object heartBeatWriteLock = new object();
+        public bool IsDisposedHeartBeatReadTimer { get; set; }
```
[the remark this answers is not in the document]
Comment (michaelklishin) [reply projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: I'd also suggest using `HasDisposedHeartbeatReadTimer` for name.

[the remark this answers is not in the document]
Comment (nazmul1985) [reply projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: Good suggestion. I will update that.

[the remark this answers is not in the document]
Comment (nazmul1985) [reply projects/client/RabbitMQ.Client/src/client/impl/Connection.cs]: I am experiencing the exception randomly when I call  Connection.Dispose(). 
In our test scenario we were opening and closing connection frequently and we able to recreate the unhandled exception every time but at random intervals. And our service crushes.

When I dispose the connection I found that it caught "System.ObjectDisposedException" exception at following line of Connection.cs:
_heartbeatReadTimer.Change(Heartbeat * 1000, Timeout.Infinite);

So, the explanation is when I call Connection.Dispose() then _heartbeatReadTimer object disposed by MaybeDisposeTimer() and the timer elapsed action (HeartbeatReadTimerCallback()) is still executing and when it tries to change the timer interval inside the callback method it found that the timer object is disposed.

**Pull request merged**